### PR TITLE
Add partial results and options

### PR DESCRIPTION
**Pull request by JohannesMoersch**
This lets you do the following (copied from the documentation update):

```csharp
Result<int, string> success = 100;
// or
Result<int, string> success = Result.Success(100);

Result<int, string> failure = "Failure";
// or
Result<int, string> failure = Result.Failure("Failure");
```

and

### With type inferencing when target type is known

```csharp
Option<int> success = 100;

Option<int> none = Option.None();
```

I definitely like the versions that still require you type Result.Success/Failure and Option.None (this couldn't be a property since the None method name conflicts). I'm torn on the implicit casts straight from the success/failure/value types to the Result/Option.

Thoughts?

### CODE REVIEW

**Comment by RyanMarcotte** [comment]
I'm sure this is buried in the changeset / documentation somewhere, but I assume the following works?

``` csharp
Result<int, string> MakeResult() => Result.Success(100);
Option<int> MakeOption() => 100;
```

File: doc/option.md [hunk 1]
```diff
@@ -16,6 +16,14 @@ Option<int> some = Option.Some(100);
 Option<int> none = Option.None<int>();
 ```
 
+### With type inferencing when target type is known
+
+```csharp
+Option<int> success = 100;
```
Comment (RyanMarcotte) [inline doc/option.md:22]: I've reached out to my team regarding their understanding of the proposed documentation, will report back

**Comment by RyanMarcotte** [comment]
Addresses #73 and #110 ?

**Comment by RyanMarcotte** [comment]
Will you be publishing a prerelease version of the package first for people to play around with?

File: doc/option.md (hunk 1 above)
Comment (RyanMarcotte) [reply doc/option.md:22]: Documentation is good as-is!

**Review by RyanMarcotte — APPROVED** [review]

**Comment by JohannesMoersch** [comment]
Yup, your examples will work with the new code. It should address both of those issues as well.

Do you think there's value in a pre-release package before pushing to main? The changes are non-breaking.

**Comment by RyanMarcotte** [comment]
>Do you think there's value in a pre-release package before pushing to main? **The changes are non-breaking.**

Given the highlighted bit, probably not.  Release away!

**Comment by JohannesMoersch** [comment]
Updated documentation:

```csharp
Result<int, string> success = 100;
Result<int, string> success = Result.Success(100);

Result<int, string> failure = "Failure";
Result<int, string> failure = Result.Failure("Failure");

// or as a return from a method
Result<int, string> MakeSuccess() => 100;
Result<int, string> MakeSuccess() => Result.Success(100);
Result<int, string> MakeFailure() => "Failure";
Result<int, string> MakeFailure() => Result.Failure("Failure");
```

```csharp
Option<int> success = 100;
Option<int> none = Option.None();

// or as a return from a method
Option<int> MakeSome() => 100;
Result<int> MakeNone() => Option.None();
```

**Comment by RyanMarcotte** [comment]
Looks good!

**Review by deimors — COMMENTED** [review]
Hmm, what happens with:
```
Result<string, string> foo = "bar";
```
Does it result in a compile time error, or an arbitrary (or declaration order) selection? I suppose I'd guess the former 🤔

**Review by michaelr-iq — APPROVED** [review]

**Comment by RyanMarcotte** [comment]
It's a compiler error for that scenario, yes

image

... which I think is fine?  I realize it's possible for `TSuccess` = `TFailure`, but how often do people actually do that?

**Comment by deimors** [comment]
Yea, I don't think that's a huge problem, so long as `Result<string, string> foo = Result.Success("bar");` still works.

For a domain with well modeled errors I agree that it's not likely to come up, but for utility methods returning a simple string error then `Result<string, string>` or `Result<T, string>` (where `T` could be `string`) isn't too uncommon.

**Comment by JohannesMoersch** [comment]
I was toying with the idea of supporting async partial results to allow the following:

```csharp
from value in Result.SuccessAsync(SomeAsyncMethod())
```

This is because `let` does not allow async, so you have to use `from` with a success result. However, because it is not possible to add an implicit typecast from `Task<PartialResult.Success<T>>` to `Task<Result<T, F>>` I don't think I will. The only way this could add value (outside of the query syntax scenario) is if I added extension methods to allow the following:

```csharp
var success = Result.SuccessAsync(SomeAsyncMethod()).ToResult<F>();
```

I'm unconvinced by the value of this so I think I will merge the PR as is.

**Comment by RyanMarcotte** [comment]
Can't say I've ever used `Result.SuccessAsync` because the vast majority of my `async` methods are implemented as methods returning `Task<Result<TSuccess, TFailure>>` already.  I'm a fan of merging these changes as-is.  Nice work!

**Review by deimors — APPROVED** [review]
👍

**Pull request merged**